### ISSUE TITLE
Feature: Improve observability around VM restarts

### ISSUE THREAD

**Issue by sharnoff** [neondatabase/autoscaling#449]
First issue with the new feature template :)

## Problem description / Motivation

There's broadly two areas of motivation here:

First: We don't have any high-level observability into the _rate of restarts and kernel panics, cluster-wide_. Sometimes we get hints of this, because failure to restart on kernel panic (will be fixed by #448) and our ongoing communications means they show up in the metrics for _other_ systems, but broadly our observability here currently comes from customer support tickets.

Second: It's hard to know if/when a _particular_ VM restarted (or had a kernel panic). Currently we have to dig through the logs, and that only _sometimes_ gives us the relevant information (c.f. [[slack link]](https://neondb.slack.com/archives/C04USJQNLD6/p1690491659926209), where a VM's logs suddenly cut out, without any shutdown or things of that nature). This is too much effort to be scalable.

## Feature idea(s) / DoD

We should have a dashboard that shows:

1. Per k8s cluster, total rate of restarts / kernel panics across all VMs; and
2. per-VM, when it's restarted / kernel panicked

And it would be _good_ to tell what the cause of a restart was, although that's significantly more difficult, and _should_ already be available via logs / k8s events. It may also be worthwhile to have a "restart count" field in the VM object's spec (although that could make idempotency hard).

There's also these _other_ types of restart to consider:

* QEMU process exits unexpectedly / killed
* Pod OOMkilled / failed / etc.

AFAIK we haven't yet seen them, but... maybe we haven't seen them because we don't have observability there?

## Implementation ideas

Related to / in addition to #448, we could use the technique outlined in answers to [this stackexchange question](https://unix.stackexchange.com/questions/443017/can-i-make-qemu-exit-with-failure-on-kernel-panic)[^stackexchange-question] to get QEMU to exit when there's a kernel panic. It seems like either way it's hard to distinguish between "normal shutdown" and "kernel panic" — maybe that's fine? The same logic in the runner to handle "QEMU exited when we weren't expecting it to" could also handle "QEMU got killed, etc". The main challenge as I see it is how to actually expose that metric — probably: how to get it to the controller? If we expose it in the pod, we'll miss kernel panics if the VM is shut down before the metrics are scraped.

In the VM controller, we also want to track the times we create new runners when the (old) runner pod becomes e.g. "succeeded" or "failed" without the VM being deleted in some metric. It could be the same as from the runner pod(s), or different.

We should also carefully consider whether we want these metrics to be per-pod or global. Or, maybe we have both, so that aggregate queries are efficient, but it's possible to look at the history for a particular VM. See also: https://neondb.slack.com/archives/C03F5SM1N02/p1690668679492609?thread_ts=1690668200.116389&cid=C03F5SM1N02

[^stackexchange-question]: NB: One of the answers says that pvpanic is only available on x86. I'm not sure that's still true: https://bugzilla.redhat.com/show_bug.cgi?id=1961178 (see "fixed in version: kernel-5.13.0...") — the question is then whether QEMU supports it.

**Comment by sharnoff** [comment]
Since this issue was opened:

- Kernel panics have significantly subsided
- New PR: https://github.com/neondatabase/autoscaling/pull/754

I also recently learned about [k8s termination messages](https://kubernetes.io/docs/tasks/debug/debug-application/determine-reason-pod-failure/#writing-and-reading-a-termination-message), which we definitely could use to provide extra information about the reason for termination (e.g. SIGTERM [polite pod termination request, like with cluster-autoscaler eviction], qemu unexpectedly exiting, container-mgr failing [from #738], OOMKilled, etc.)

IMO this should be fairly straightforward after #754. This is somewhat similar to #759. Not sure whether it's worth including in neondatabase/company_projects#187.

**Comment by shayanh** [comment]
k8s termination messages define a termination messages for a container inside a pod. Since we delete runner pod in case of any failure, basically writing anything there won't be useful unless we define a similar field in the VM status to track last state. I open a new issue for that. Regarding this one, I'll be adding cluster-wide and per-VM restart metrics.

**Comment by shayanh** [comment]
For a specific VM, the reason for a restart can be tracked also through `.Status.Conditions`.
